typed-store/src/rocks/mod.rs

```diff
@@ -405,15 +447,38 @@ where
     }
 }
 
+// This is used internally as a merge operator, see:
+// https://github.com/facebook/rocksdb/wiki/Merge-Operator
+pub(crate) fn non_conflicting_insert_merge(
+    _new_key: &[u8],
+    existing_val: Option<&[u8]>,
+    operands: &MergeOperands,
+) -> Option<Vec<u8>> {
+    let nops =
+        operands.iter().map(|v| v.len()).sum::<usize>() + existing_val.map(|_| 1).unwrap_or(0);
+    let mut result: Vec<u8> = Vec::with_capacity(nops);
+    if let Some(v) = existing_val {
+        result.extend(v.iter());
```

Comment:
There's no need to compute nops and allocate a result if there is an existing val, it's wasted.  Also not sure how fast v.iter() is, but it's probably optimized.  However I would do this which should be faster and simpler:

```rust
if let Some(v) = existing_val {
    Some(v.clone())
} else if let Some(op) = operands.into_iter().next() {
    Some(op.clone())
}
```

Reply:
This is not a total map (`existing_val` and `operands` can both be `None`), and `v`,  `op` are slices, which require a `to_vec()` to allocate besides. But I took some of the feedback into account nonetheless.